vertx-web/src/main/java/io/vertx/ext/web/impl/SignatureImpl.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2023 Red Hat, Inc.
+ *
+ *  All rights reserved. This program and the accompanying materials
+ *  are made available under the terms of the Eclipse Public License v1.0
+ *  and Apache License v2.0 which accompanies this distribution.
+ *
+ *  The Eclipse Public License is available at
+ *  http://www.eclipse.org/legal/epl-v10.html
+ *
+ *  The Apache License v2.0 is available at
+ *  http://www.opensource.org/licenses/apache2.0.php
+ *
+ *  You may elect to redistribute this code under either of these licenses.
+ */
+
+package io.vertx.ext.web.impl;
+
+import io.vertx.core.impl.logging.Logger;
+import io.vertx.core.impl.logging.LoggerFactory;
+import io.vertx.ext.web.Signature;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidKeyException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+
+import static io.vertx.ext.auth.impl.Codec.base64UrlEncode;
+
+public class SignatureImpl implements Signature {
+  private static final Logger LOG = LoggerFactory.getLogger(SignatureImpl.class);
+  private static final String algorithm = "HmacSHA256";
+
+  private final Mac mac;
+
+  public SignatureImpl(final String secret) {
+    try {
+      if (secret.length() <= 8) {
+        LOG.warn("Signing secret is very short (<= 8 bytes)");
+      }
+      mac = Mac.getInstance(algorithm);
+      mac.init(new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), algorithm));
+    } catch (NoSuchAlgorithmException | InvalidKeyException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public String sign(String data) {
+    final String signature;
+    synchronized (mac) {
+      signature = base64UrlEncode(mac.doFinal(data.getBytes(StandardCharsets.US_ASCII)));
+    }
+    return data + "." + signature;
+  }
+
+  @Override
+  public boolean verify(String signedData) {
+    String[] parts = signedData.split("\\.");
+    if (parts.length < 2) {
+      // There is no signature on this data
+      return false;
+    }
+
+    final String signature = parts[parts.length - 1];
+    final String[] dataParts = Arrays.copyOfRange(parts,0, parts.length - 1);
+    final String data = String.join(".", dataParts);
+    final String calculatedSignature;
+
+    synchronized (mac) {
```

Comment:
Out of curiosity, why do we need to synchronize on `mac`?

Reply:
I'll be honest with you I took it from the code that was in the CSRF handler and like you wasn't 100% sure why it was there, but assumed someone must have put it there for a good reason so kept it. 

Happy to defer to your suggestion here